src/metabase/lib/join.cljc

```diff
@@ -584,3 +584,49 @@
      (when-let [fk-col (fk-column-for query stage-number pk-col)]
        (lib.common/->op-arg
         (lib.filter/filter-clause (equals-join-condition-operator-definition) fk-col pk-col))))))
+
+(defn- replace-join-alias
+  [a-join old-name new-name]
+  (mbql.u.match/replace a-join
+    (field :guard (fn [field-clause]
+                    (and (field-clause? field-clause)
+                         (= (current-join-alias field-clause) old-name))))
+    (with-join-alias field new-name)))
+
+(defn- rename-join-in-stage
+  [stage old-name-or-index new-name]
+  (let [the-joins (:joins stage)
+        idx       (if (integer? old-name-or-index)
+                    (when (< -1 old-name-or-index (count the-joins))
+                      old-name-or-index)
+                    (some (fn [[idx a-join]]
+                            (when (= (:alias a-join) old-name-or-index)
+                              idx))
+                          (m/indexed the-joins)))
+        old-name  (get-in the-joins [idx :alias])]
+    (if (and idx (not= old-name new-name))
+      (let [unique-name-fn (lib.util/unique-name-generator)
+            _              (run! unique-name-fn (map :alias the-joins))
+            unique-name    (unique-name-fn new-name)]
+        (-> stage
+            (assoc-in [:joins idx :alias] unique-name)
+            (update-in [:joins idx] replace-join-alias old-name unique-name)))
```

Comment:
I thought join field references can exist with a `:join-alias` elsewhere within the stage?

e.g.
```clojure
{:joins [{... :alias "foo"}]
 :filters [[:= {...} [:field {:join-alias "foo"} 123] 1]]}
 ```